### PR TITLE
[DEVX-219]: Model Training

### PR DESCRIPTION
**Pull request by sanjaychelliah**
# What

- Integrating Model Training in SDK
- Adding support functions for training

# Features

#### Model Training
```python
# Note: CLARIFAI_PAT must be set as env variable.
from clarifai.client.app import App
from clarifai.client.model import Model

"""
Create model with trainable model_type
"""
app = App(user_id="user_id", app_id="app_id")
model = app.create_model(model_id="model_id", model_type_id="visual-classifier")
               (or)
model = Model('url')

"""
List training templates for the model_type
"""
templates = model.list_training_templates()
print(templates)

"""
Get parameters for the model.
"""
params = model.get_params(template='classification_basemodel_v1', yaml_file='model_params.yaml'
)

"""
Update the model params yaml and pass it to model.train()
"""
model_version_id = model.train('model_params.yaml')

"""
Training status and saving logs
"""
status = model.training_status(version_id=model_version_id,training_logs=True)
print(status)
```

#### Model Training support functionalities

```python
# Note: CLARIFAI_PAT must be set as env variable.
from clarifai.client.app import App
from clarifai.client.model import Model

"""
show trainable model types
"""
app = App(user_id="user_id", app_id="app_id")
print(app.list_trainable_model_types())

"""
Show individual param information
"""
params = model.get_params(template='classification_basemodel_v1', yaml_file='model_params.yaml')
print(model.get_param_info(param = 'dataset_id))

"""
Update specific params for the model
"""
model.update_params({'template': 'xxx'})
model.train()
```

#### Dataset functionalities added

- dataset.create_version()
- dataset.list_versions()
- dataset.delete_version()

## TODO

- Pytest tests in the upcoming PR
- More examples in [examples](https://github.com/Clarifai/examples/tree/main) repo

### CODE REVIEW

**Review by ackizilkale — APPROVED** [review]
looks good to me.

File: clarifai/client/model.py [hunk 1]
```diff
@@ -47,10 +57,243 @@ def __init__(self,
...
+      )
+
+    all_keys = [key for key in self.training_params.keys()] + [
+        key for key in self.training_params.values() if isinstance(key, dict) for key in key
+    ]
+    if not set(kwargs.keys()).issubset(all_keys):
+      raise UserError("Invalid params")
+
+    for key, value in kwargs.items():
+      find_and_replace_key(self.training_params, key, value)
```
Comment (sainivedh) [inline clarifai/client/model.py:158]: Could you please add some comments for this

File: clarifai/client/model.py [hunk 2]
```diff
@@ -47,10 +57,243 @@ def __init__(self,
...
+    """Updates the model params for the model.
+
+    Args:
+        **kwargs: model params to update.
+
+    Example:
+        >>> from clarifai.client.model import Model
+        >>> model = Model(model_id='model_id', user_id='user_id', app_id='app_id')
+        >>> model_params = model.get_params(template='template', yaml_file='model_params.yaml')
+        >>> model.update_params({param1:value1, param2:value2})
```
Comment (sainivedh) [inline clarifai/client/model.py:142]:
```suggestion
        >>> model.update_params({batch_size:8, dataset_version:v2})
```
I think having some example params helps

File: clarifai/client/model.py [hunk 3]
```diff
@@ -47,10 +57,243 @@ def __init__(self,
...
+        model_versions=[resources_pb2.ModelVersion(**train_dict)])
+    response = self._grpc_request(self.STUB.PostModelVersions, request)
+    if response.status.code != status_code_pb2.SUCCESS:
+      raise Exception(response.status)
+    self.logger.info("\nModel Training Started\n%s", response.status)
+
+    return response.model.model_version.id
+
+  def training_status(self, version_id: str, training_logs: bool = False) -> Dict[str, str]:
+    """Returns the training status for the model version.
```
Comment (sainivedh) [inline clarifai/client/model.py:240]:
```suggestion
    """Get the training status for the model version also stores training logs
```

File: clarifai/client/model.py [hunk 4]
```diff
@@ -47,10 +57,243 @@ def __init__(self,
...
+    if response.status.code != status_code_pb2.SUCCESS:
+      raise Exception(response.status)
+
+    if training_logs:
+      try:
+        if response.model_version.train_log:
+          log_response = requests.get(response.model_version.train_log)
+          log_response.raise_for_status()  # Check for any HTTP errors
+          with open(version_id + '.log', 'wb') as file:
+            file.write(log_response.content)
```
Comment (sainivedh) [inline clarifai/client/model.py:269]:
```suggestion
            for chunk in log_response.iter_content(chunk_size=4096):  # 4KB
                  file.write(chunk)
```
scalable way to save large logs that may not fit into memory

File: clarifai/utils/model_train.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+          del modeltypefield['placeholder']
+          return modeltypefield
+        if modeltypefield['path'].split('.')[-1] == "template":
+          for modeltypeenum in modeltypefield['modelTypeEnumOptions']:
+            if modeltypeenum['id'] == template:
+              for modeltypeenumfield in modeltypeenum['modelTypeFields']:
+                if modeltypeenumfield['path'].split('.')[-1] == param:
+                  modeltypeenumfield['param'] = modeltypeenumfield.pop('path').split('.')[-1]
+                  del modeltypeenumfield['placeholder']
+                  return modeltypeenumfield
```
Comment (sainivedh) [inline clarifai/utils/model_train.py:140]: Some comments could be helpful to maintain in future

File: clarifai/utils/model_train.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
+from typing import Any, Dict, List
+
+from clarifai_grpc.grpc.api import resources_pb2
+from clarifai_grpc.grpc.api.service_pb2 import MultiModelTypeResponse
+from google.protobuf.json_format import MessageToDict
+from google.protobuf.struct_pb2 import Struct
+
+
+def response_to_templates(response: MultiModelTypeResponse, model_type_id: str) -> List[str]:
```
Comment (sainivedh) [inline clarifai/utils/model_train.py:9]:
```suggestion
def _response_to_templates(response: MultiModelTypeResponse, model_type_id: str) -> List[str]:
```
Shall we mark all these function as internal use ? wdyt

**Review by sainivedh — COMMENTED** [review]
Great Work. Added some comments

File: clarifai/client/model.py [hunk 7]
```diff
@@ -15,6 +17,14 @@
...
 from clarifai.utils.logging import get_logger
 from clarifai.utils.misc import BackoffIterator
+from clarifai.utils.model_train import (find_and_replace_key, params_parser,
+                                        response_to_model_params, response_to_param_info,
+                                        response_to_templates)
+
+TRAINABLE_MODEL_TYPES = [
+    'visual-classifier', 'visual-detector', 'visual-segmenter', 'visual-anomaly-heatmap',
+    'visual-embedder', 'clusterer', 'text-classifier', 'embedding-classifier', 'text-to-text'
+]
```
[the remark this answers is not in the document]
Comment (sanjaychelliah) [reply clarifai/client/model.py:27]: Yeah 👍

File: clarifai/utils/model_train.py (hunk 6 above)
Comment (sanjaychelliah) [reply clarifai/utils/model_train.py:9]: The functions are stored in internal utils and we are following the same structure in [Workflows](https://github.com/Clarifai/clarifai-python/blob/d91cd2bfa2ece8813cb1448fc802289be1181a3e/clarifai/client/app.py#L22). So I think this is fine.

File: clarifai/client/model.py (hunk 4 above)
Comment (sanjaychelliah) [reply clarifai/client/model.py:269]: Sure

File: clarifai/client/model.py [hunk 8]
```diff
@@ -47,10 +57,243 @@ def __init__(self,
...
+    request = service_pb2.ListModelTypesRequest(user_app_id=self.user_app_id,)
+    response = self._grpc_request(self.STUB.ListModelTypes, request)
+    if response.status.code != status_code_pb2.SUCCESS:
+      raise Exception(response.status)
+    templates = response_to_templates(
+        response=response, model_type_id=self.model_info.model_type_id)
+
+    return templates
+
+  def get_params(self, template: str = None, yaml_file: str = 'params.yaml') -> Dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (sanjaychelliah) [reply clarifai/client/model.py:88]: Yeah seems relevant 👍

File: clarifai/client/model.py (hunk 1 above)
Comment (sanjaychelliah) [reply clarifai/client/model.py:158]: Sure

File: clarifai/utils/model_train.py (hunk 5 above)
Comment (sanjaychelliah) [reply clarifai/utils/model_train.py:140]: Sure. We can add comments as `parsing the API response Multimodeltypes.` But every loop detail, not sure. Thats why kept the iter variable names clearly.

File: clarifai/client/model.py [hunk 9]
```diff
@@ -70,7 +66,7 @@ def list_training_templates(self) -> List[str]:
     Example:
         >>> from clarifai.client.model import Model
         >>> model = Model(model_id='model_id', user_id='user_id', app_id='app_id')
-        >>> model.list_training_templates()
+        >>> print(odel.list_training_templates())
```
Comment (sainivedh) [inline clarifai/client/model.py:69]: **model**

File: README.md [hunk 10]
```diff
@@ -219,8 +219,7 @@ print(templates)
 """
 Get parameters for the model.
 """
-params = model.get_params(template='classification_basemodel_v1', yaml_file='model_params.yaml'
-)
+params = model.get_params(template='classification_basemodel_v1', yaml_file='model_params.yaml')
```
Comment (sainivedh) [inline README.md:222]: `save_to`

**Review by sainivedh — APPROVED** [review]
lgtm!

**Pull request merged**